security/pkg/pki/ca/ca.go

```diff
@@ -317,9 +329,244 @@ func (ca *IstioCA) SignWithCertChain(csrPEM []byte, subjectIDs []string, ttl tim
 
 // GetCAKeyCertBundle returns the KeyCertBundle for the CA.
 func (ca *IstioCA) GetCAKeyCertBundle() util.KeyCertBundle {
+	ca.mutex.RLock()
+	defer ca.mutex.RUnlock()
 	return ca.keyCertBundle
 }
 
+// setCAKeyCertBundle sets KeyCertBundle to the CA.
+func (ca *IstioCA) setCAKeyCertBundle(newBundle util.KeyCertBundle) {
+	ca.mutex.Lock()
+	defer ca.mutex.Unlock()
+	ca.keyCertBundle = newBundle
+}
+
+// SelfSignedCARootCertRotationConfig configs the automatic root certificate
+// rotation for self-signed CA instance.
+type SelfSignedCARootCertRotationConfig struct {
+	CheckInterval       time.Duration
+	CaCertTTL           time.Duration
+	RetryInterval       time.Duration
+	GracePeriodRatio    float32
+	Client              corev1.CoreV1Interface
+	CaStorageNamespace  string
+	DualUse             bool
+	ReadSigningCertOnly bool
+	Org                 string
+	RootCertFile        string
+	Metrics             monitoring.MonitoringMetrics
+}
+
+// RotateRootCert refreshes root certs and updates config map accordingly.
+func (ca *IstioCA) RotateRootCert(config *SelfSignedCARootCertRotationConfig) {
+	ticker := time.NewTicker(config.CheckInterval)
+	for {
+		select {
+		case <-ticker.C:
+			ca.checkAndRotateRootCert(config)
+		case _, ok := <-ca.StopRotateJob:
+			if !ok {
+				log.Info("Self-signed CA is shutting down, stop root cert rotation job")
+				if ticker != nil {
+					ticker.Stop()
+				}
+				return
+			}
+		}
+	}
+}
+
+type RootUpgradeStatus int
+
+const (
+	UpgradeSkip    RootUpgradeStatus = 0
+	UpgradeSuccess RootUpgradeStatus = 1
+	UpgradeFailure RootUpgradeStatus = 2
+)
+
+// checkAndRotateRootCert decides whether root cert should be refreshed, and rotates
+// root cert for self-signed Citadel.
+func (ca *IstioCA) checkAndRotateRootCert(config *SelfSignedCARootCertRotationConfig) {
+	caSecret, scrtErr := ca.loadCASecretWithRetry(config)
+
+	if config.ReadSigningCertOnly {
+		status := ca.checkAndRotateRootCertForReadOnlyCitadel(config, caSecret, scrtErr)
+		if status == UpgradeSuccess {
+			config.Metrics.RootUpgradeSuccess.Increment()
+		}
+		if status == UpgradeFailure {
+			config.Metrics.RootUpgradeErrors.Increment()
+		}
+		return
+	}
+
+	status := ca.checkAndRotateRootCertForSigningCertCitadel(config, caSecret, scrtErr)
+	if status == UpgradeSuccess {
+		config.Metrics.RootUpgradeSuccess.Increment()
+	}
+	if status == UpgradeFailure {
+		config.Metrics.RootUpgradeErrors.Increment()
+	}
+	return
+}
+
+// loadCASecretWithRetry lets the self-signed Citadel read CA secret with retries until
+// timeout.
+func (cs *IstioCA) loadCASecretWithRetry(config *SelfSignedCARootCertRotationConfig) (*v1.Secret, error) {
```

Comment:
Here you use `cs` the rest is `ca`.

Do you have to write this func yourself? Don't we need this functionality in Citadel start up before already?